examples/dann.py

```diff
@@ -0,0 +1,307 @@
+X_test = np.concatenate([X_test, X_test, X_test], axis=1)
+
+X_train = X_train.astype('float32')
+X_test = X_test.astype('float32')
+X_train /= 255
+X_test /= 255
+
+XT_train = XT_train.astype('float32')
+XT_test = XT_test.astype('float32')
+XT_train /= 255
+XT_test /= 255
+
+domain_labels = np.vstack([np.tile([0, 1], [batch_size / 2, 1]),
+                           np.tile([1., 0.], [batch_size / 2, 1])])
+
+# Created mixed dataset for TSNE visualization
+num_test = 500
+combined_test_imgs = np.vstack([X_test[:num_test], XT_test[:num_test]])
+combined_test_labels = np.vstack([y_test[:num_test], y_test[:num_test]])
+combined_test_domain = np.vstack([np.tile([1., 0.], [num_test, 1]),
+                                 np.tile([0., 1.], [num_test, 1])])
+
+
+class DANNBuilder(object):
+    def __init__(self):
+        self.model = None
+        self.net = None
+        self.domain_invariant_features = None
+        self.grl = None
+        self.opt = SGD()
+
+    def _build_feature_extractor(self, model_input):
+        '''Build segment of net for feature extraction.'''
+        net = Convolution2D(nb_filters, nb_conv, nb_conv,
+                            border_mode='valid',
+                            activation='relu')(model_input)
+        net = Convolution2D(nb_filters, nb_conv, nb_conv,
+                            activation='relu')(net)
+        net = MaxPooling2D(pool_size=(nb_pool, nb_pool))(net)
+        net = Dropout(0.5)(net)
+        net = Flatten()(net)
+        self.domain_invariant_features = net
+        return net
+
+    def _build_classifier(self, model_input):
+        net = Dense(128, activation='relu')(model_input)
+        net = Dropout(0.5)(net)
+        net = Dense(nb_classes, activation='softmax',
+                    name='classifier_output')(net)
+        return net
+
+    def build_source_model(self, main_input, plot_model=False):
+        net = self._build_feature_extractor(main_input)
+        net = self._build_classifier(net)
+        model = Model(input=main_input, output=net)
+        if plot_model:
+            plot(model, show_shapes=True)
+        model.compile(loss={'classifier_output': 'categorical_crossentropy'},
+                      optimizer=self.opt, metrics=['accuracy'])
+        return model
+
+    def build_dann_model(self, main_input, hp_lambda, plot_model=False):
+        net = self._build_feature_extractor(main_input)
+        self.grl = GradientReversal()
+        branch = self.grl(net, hp_lambda)
+        branch = Dense(128, activation='relu')(branch)
+        branch = Dropout(0.1)(branch)
+        branch = Dense(2, activation='softmax', name='domain_output')(branch)
+
+        # When building DANN model, route first half of batch (source examples)
+        # to domain classifier, and route full batch (half source, half target)
+        # to the domain classifier.
+        net = Lambda(lambda x: K.switch(K.learning_phase(), x[:int(batch_size / 2), :], x, lazy=True),
```

Comment:
You're welcome! Hope it will help to merge this PR with main branch, I believe, that gradient reversal layer and domain adaptation method, that use it, can be quite useful for the community.